# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [86]:
def add_two(x):
    return x + 2

print(add_two(4))

# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(4))

6
6


#### Saving to a Variable

In [87]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [88]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [89]:
def apply_func(f, name):
    return f(name)

print(apply_func(str.upper, 'mark'))

def multiply(f,num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4) # lambda 4: 4 * 4

MARK


16

In [90]:
apply_func(lambda a_string: a_string[-1], 'MarkyMark')

'k'

#### Returning a Lambda from a Function

In [91]:
#Regular defined function
def multiply_test(num):
    a = 5
    def inner(c, d):
        return (a*num) - (c*d)
    return inner




test = multiply_test(10)
print(test)
print(type(test))


<function multiply_test.<locals>.inner at 0x000002D51A0DFA60>
<class 'function'>


In [92]:
test(3, 5)

35

In [93]:
def return_lambda(num):
    a = 5
    return lambda c,d: (num*a) - (c*d)

lamb_test = return_lambda(15)

print(lamb_test)
print(type(lamb_test))


<function return_lambda.<locals>.<lambda> at 0x000002D51A594160>
<class 'function'>


In [94]:
lamb_test(4,7)

47

#### If Statements within Lambdas

In [95]:
def do_something(num):
    if num > 10:
        return num * 2
    else:
        return num+2
    

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [97]:
f = lambda num: num**3
f(10)

1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [98]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Using Lambda's with Map

In [99]:
# map(function,  iterable -> list,tuple,set,etc)

def squared_minus_two(a):
    return a**2 - 2

numbers = [2, 5, 6, 8, 2, 6, 2, 7, 8, 3, 7]

mapped_nums = map(squared_minus_two, numbers)
mapped_nums_list = list(mapped_nums)
mapped_nums_list

[2, 23, 34, 62, 2, 34, 2, 47, 62, 7, 47]

In [100]:
nums = [2, 5, 4, 7, 9]
more_nums = [3, 4, 6, 1]

def some_func(a, b):
    if a > b:
        return (a, b**2)
    else:
        return (a**2, b)
    

list(map(some_func, nums, more_nums))

[(4, 3), (5, 16), (16, 6), (7, 1)]

In [101]:
mapped_squares_list = list(map(lambda x: x**2 -2 , numbers))
print(mapped_squares_list)

[2, 23, 34, 62, 2, 34, 2, 47, 62, 7, 47]


In [102]:
list(map(lambda x,y: (x,y**2) if x > y else (x**2, y), nums, more_nums))

[(4, 3), (5, 16), (16, 6), (7, 1)]

In [103]:
# rng = [1,4,2,5,6]
# mult_two_minus_one = list(map(lambda x: x*2 -1, rng))
# print(mult_two_minus_one)

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if its odd</p>

In [104]:
rng = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

even_or_odd = list(map(lambda x: x*2 if x %2 == 0 else x/2, rng))
print(even_or_odd)

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [105]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [106]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def starts_with_a(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
a_names_mapped = list(map(starts_with_a, names))
print(a_names_mapped)

a_names_filtered = list(filter(starts_with_a, names))
print(a_names_filtered)

[False, True, False, False, True]
['Andy', 'Angelica']


In [107]:
name = "Alex"
name[0].lower() == 'a'

True

#### Using Lambda's with Filter()

In [108]:
a_names = list(filter(lambda n: True if n[0].lower() == 'a' else False , names))
print(a_names)

a_names = list(filter(lambda n: n[0].lower() == 'a', names))
print(a_names)

['Andy', 'Angelica']
['Andy', 'Angelica']


In [109]:
def remainder(x):
    return x % 3 


list(filter(remainder, [1, 2, 3, 4, 5, 6, 7, 8, 9]))

[1, 2, 4, 5, 7, 8]

In [110]:
list(map(remainder, [1, 2, 3, 4, 5, 6, 7, 8, 9]))

[1, 2, 0, 1, 2, 0, 1, 2, 0]

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [111]:
import statistics
numbers = [15, 3, 57, 8, 13, 56, 40, 83, 83, 90]

In [112]:
mean_of_nums = statistics.mean(numbers)
mean_of_nums

44.8

In [113]:
below_mean = list(filter(lambda x: x < mean_of_nums, numbers))
below_mean

[15, 3, 8, 13, 40]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [114]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [115]:
list_1 = [2, 4, 6, 8 , 10]

def add_nums(num1, num2):
    print('=========')
    print(f"num1: {num1}")
    print(f"num2: {num2}")
    return num1+num2


result_add = reduce(add_nums, list_1)
result_add

num1: 2
num2: 4
num1: 6
num2: 6
num1: 12
num2: 8
num1: 20
num2: 10


30

In [116]:
def sub_nums(num1, num2):
    print('=========')
    print(f"num1: {num1}")
    print(f"num2: {num2}")
    return num1 - num2

result_sub = reduce(sub_nums, list_1)
result_sub

num1: 2
num2: 4
num1: -2
num2: 6
num1: -8
num2: 8
num1: -16
num2: 10


-26

#### Using Lambda's with Reduce()

In [117]:
result_lamb = reduce(lambda num1, num2: num1 + num2, list_1)
print(result_lamb)

30


In [118]:
#using a for loop instead

total = 0
for num in list_1:
    total += num
print(total)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [119]:
list_1 = [2, 4, 6, 8 ,10]

mult_lamb = reduce(lambda num1, num2: num1 * num2 , list_1)
mult_lamb

3840

In [120]:
total = 1
for num in list_1:
    total *= num
print(total)

3840


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [121]:
def full_name(first, last):
    return first+" "+ last


def greet_person(first, last, day_of_week):
    person_full_name = full_name(first, last)
    print(f'Hello {person_full_name}, it sure is beautiful {day_of_week}')

greet_person('Mark', 'Torres', 'Tuesday')

Hello Mark Torres, it sure is beautiful Tuesday


#### Implementing a Base Case

In [124]:
def add_nums(num):
    #Set base case for recursive function
    if num <= 1:
        print(f'add_nums({num}) = {num}')
        return num
    else:
        print(f'add_nums({num}) =  {num} + add_nums({num - 1})')
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) =  5 + add_nums(4)
add_nums(4) =  4 + add_nums(3)
add_nums(3) =  3 + add_nums(2)
add_nums(2) =  2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [130]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(n):
    if n<=1:
        print(f'{n}! = 1')
        return 1
    else:
        print(f'{n}! = {n} * {n-1}!')
        return n *factorial(n-1)
    
factorial(5)

5! = 5 * 4!
4! = 4 * 3!
3! = 3 * 2!
2! = 2 * 1!
1! = 1


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [133]:
def sub_nums(num):
    #Set base case for recursive function
    if num <= 1:
        print(f'sub_nums({num}) = {num}')
        return num
    else:
        print(f'sub_nums({num}) =  {num} + sub_nums({num - 1})')
        return num - sub_nums(num - 1)
    
sub_nums(10)

sub_nums(10) =  10 + sub_nums(9)
sub_nums(9) =  9 + sub_nums(8)
sub_nums(8) =  8 + sub_nums(7)
sub_nums(7) =  7 + sub_nums(6)
sub_nums(6) =  6 + sub_nums(5)
sub_nums(5) =  5 + sub_nums(4)
sub_nums(4) =  4 + sub_nums(3)
sub_nums(3) =  3 + sub_nums(2)
sub_nums(2) =  2 + sub_nums(1)
sub_nums(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [136]:
a = map(len, ['Brian','Tatyana', 'Ryan'])
a[2]

TypeError: 'map' object is not subscriptable

In [137]:
for i in a:
    print(i)

5
7
4



#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def my_range(start,stop,step=1):
    

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
#FOR REFERENCE
# a_names = list(filter(lambda n: True if n[0].lower() == 'a' else False , names))
# print(a_names)

# a_names = list(filter(lambda n: n[0].lower() == 'a', names))
# print(a_names)

In [147]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filtered_list = list(filter(lambda x: x.strip(), places))

print(filtered_list)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [172]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key = lambda name: name.rsplit()[len(name.split())-1].lower())
author

['Victor aNisimov',
 'Gary A.J. Bernstein',
 'Joel Carter',
 'Andrew P. Garfield',
 'David hassELHOFF']

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [166]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

places_farhenheit = list(map(lambda tup: (tup[0], tup[1] * 9/5 + 32), places))
places_farhenheit

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [177]:
def fib(n):
    if n <= 1:
        return n
    else:
        print(f'Iteration{n}: {n}')
        return(fib(n-1) + fib(n-2))

              
fib(5)
#¯\_(ツ)_/¯

Iteration5: 5
Iteration4: 4
Iteration3: 3
Iteration2: 2
Iteration2: 2
Iteration3: 3
Iteration2: 2


5